ui/lib/apps/OptimizerTrace/index.tsx

```diff
@@ -0,0 +1,243 @@
+import React, { useCallback, useState } from 'react'
+import { HashRouter as Router, Routes, Route } from 'react-router-dom'
+import { Button, Upload, Space } from 'antd'
+import { UploadOutlined, ArrowRightOutlined } from '@ant-design/icons'
+
+import { Card, Toolbar, Root } from '@lib/components'
+
+import styles from './index.module.less'
+import LogicalOperatorTree, {
+  LogicalOperatorNode,
+} from './components/LogicalOperatorTree'
+import PhysicalOperatorTree, {
+  PhysicalOperatorNode,
+} from './components/PhysicalOperatorTree'
+
+export default function OptimizeTraceApp() {
+  return (
+    <Root>
+      <Router>
+        <Routes>
+          <Route path="/optimizer_trace" element={<OptimizerTrace />} />
+        </Routes>
+      </Router>
+    </Root>
+  )
+}
+
+interface OptimizerData {
+  logical: {
+    final: LogicalOperatorNode[]
+    steps: LogicalOptimizeActionStep[]
+  }
+  physical: {
+    final: LogicalOperatorNode
+    selected_candidates: PhysicalOperatorNode[]
+    discarded_candidates: PhysicalOperatorNode[]
+  }
+  final: LogicalOperatorNode[]
+  isFastPlan: boolean
+}
+
+interface LogicalOptimizeActionStep {
+  index: number
+  name: string
+  before: LogicalOperatorNode[]
+  steps: {
+    id: number
+    index: number
+    action: string
+    reason: string
+    type: string
+  }[]
+}
+
+function OptimizerTrace() {
+  const [data, setData] = useState<OptimizerData | null>(null)
+
+  const handleBeforeUpload = useCallback(async (file: File) => {
+    const t = await file.text()
+    setData(JSON.parse(t))
+    return false
+  }, [])
+
+  return (
+    <div>
+      <Card noMarginBottom style={{ height: '70px' }}>
+        <Toolbar>
+          <Space>
+            <Upload
+              beforeUpload={handleBeforeUpload}
+              accept=".json"
+              maxCount={1}
+            >
+              <Button icon={<UploadOutlined />}>Select File</Button>
+            </Upload>
+          </Space>
+        </Toolbar>
+      </Card>
+
+      {data && (
```

Comment:
It will be better check the data validation, if the json fields are not those defined in interface, you may need to give a warning to users.